### DISCUSSION TITLE
Pre-alpha mission restoration

### DISCUSSION THREAD

**Discussion by mefistotelis** [swfans/syndwarsfx#147] [General]
The map from pre-alpha is buggy and unfinished, and level is is mostly automatic generation, with some behaviors, which look like they're scripted, being hard-coded into game code (ie. civilian women following a suitcase man or fight near one of the the top map corners).

There's also no objective. You just wander around until you get bored.

Still, it's an interesting exercise to make that work within SW Port.

I just made several fixes to DAT loading, and the original DAT map now has lightning which matches the look in pre-alpha engine.

Some issues in map:
* water does not wobble, only animates - but that replicates the original behavior from pre-alpha, so not sure if we should fix that
* A shallow trench near the rives is steep enough to disallow walk, dividing the map into two separate areas [fixed]
* tracing a path on the destructible bridge does not work - my pathfinding re-creation doesn't support that yet

Level file for that was created by @Moburma.
Both map and level were not added to swars-levels repo yet.

**Comment by Moburma** [comment]
Yeah, I have to admit I got bored of this once I realised the limitations. It's cool to have, but I soon realised that it would be quite a bit of effort to recreate the original without much real payoff. Not being able to walk across the map at all breaks most of the interesting stuff like the women following the man, and the other behaviours exhibited in the demo is actually quite hard to mimic without having to create loads of commands for each NPC (and you have to figure out who was doing what originally), as in the original they do all kinds of peculiar things (only attacking certain enemies, walking towards that big battle). Even the basic Wander command shows the difference in pathfinding and behaviour between versions.

Also I converted it using my crappy old script to version 17 (because I haven't supported the pre-alpha level formats in my editor yet), which means it has weird bugs, so I really need to do it again properly. I might stick it up here for people to play with if there is interest.

The other problem with the map is it can't even really be used for new levels. Even in the original demo it doesn't seem possible to go over the large road bridge, it looks to me like this was some dead end of development that was replaced with the proper roads. Also it has a weird scale, with buildings way too small for the people. Might make a nice starting point for someone to make a new map, if we ever get there, though.

**Reply by mefistotelis** [reply]
> Not being able to walk across the map at all breaks most of the interesting stuff

Ok, you persuaded me to create more complex rules for steepness. The map is now walkable.

Surprisingly, the wandering crowd didn't changed its behavior - it is still starting to wallhug around the same areas.


>  it doesn't seem possible to go over the large road bridge, it looks to me like this was some dead end of development

Oh no, that was developed.
This bridge and the similar roads are the example of "bezier roads" - you draw a bezier curve, and the editor would generate a road following that curve. Only in pre-alpha the roads generated are very close to ground.

So this is basically an early version of what is used in [New York map](http://syndicate.lubiki.pl/swars/walkthrough/swars_lev072.php)

**Reply by mefistotelis** [reply]
Hm, for the commands which the people have - they seem to lean towards that point with a big fight after a few are shot there. So comparing that to final, this looks like what the police does on some maps - walk towards places where fights happen.

**Reply by Moburma** [reply]
Ahh, so it is the same road technology, it clearly wasn't working at that point in time (all the videos from there avoid it too). I'll have another go at this level if it's walkable now. It is weird though, I guess it could be using the Guardian values or something, but like you get some combat happening right at the start of the level that doesn't seem to effect most NPCs, they just ignore each other. So it would need either lame specific kill person X commands, or a lot of sub groups for each type of NPC to get them to behave the same way. I will play around with it.

Also is the billboard textures thing the "video screen" style effect in the alpha? It's funny that this is shown off loads in the demo and then not even used once in the final game.

**Reply by mefistotelis** [reply]
> Also is the billboard textures thing the "video screen" style effect in the alpha?

I don't think I understand.
There are two effect: play FLI video, and strange "mirror" effect which is really taking the video buffer from previous frame and making it into a reflection from walls (you can even see interface items in that reflection).

The billboard FLI playback was later optimized, so that it happens only when billboard is on-screen.
Well, it was also moved to another texture - in pre-alpha it was in top left corner of texture atlas page 5, in final it's bottom center of texture atlas page 4.
Anyway, the optimization - it is implemented in a crude way, `if (thing->Subtype == BILLBOARD) then set_flag_which_continues_playback()`. So if thing of type OBJECT and subtype BILLBOARD is not on screen, the FLI does not play.
So to make the videos on pre-alpha level play in SW Port, you'd have to set the flag on each game turn.

Details for setting the flag are in the issue:
https://github.com/swfans/swars/issues/145

**Reply by Moburma** [reply]
Yeah sorry, I meant the "mirror" effect. That's not used at all in the final game, so I assume it's probably not even in the code anymore.

**Reply by mefistotelis** [reply]
I'm sure the effect would be easy to re-create. Though for me, this effect doesn't seem interesting in any way.
I don't know where the currently visible noise is coming from - maybe figuring that out would allow for an easy fix.

A real mirror effect would be harder, but more interesting I think. Though we need to be far deeper into the rewrite to think about that.

**Reply by Moburma** [reply]
Yeah, I agree, it's not worth restoring, I just wondered if it was still in the final game after all.

Anyway, I wasn't well last night and couldn't sleep, so I ended up putting in the hours to get the programming chores done needed to make a second attempt at this. 

It's going well, with a much better base, and of course the pathfinding now works!

![](https://i.ibb.co/0Q0X6RR/Desktop-Screenshot-2024-02-16-03-50-51-10.png)

Still got lots more to do, but it's coming along nicely.

**Reply by mefistotelis** [reply]
> it's not worth restoring, I just wondered if it was still in the final game after all.

Not sure. I would have to first understand where the currently seen noise is coming from.
I might investigate it - it does makes me wonder..

**Comment by mefistotelis** [comment]
Update on mirror textures:
if `GFlags & 0x02` is set within a face, then that face uses `scratch_buf1` as texture and ignores the normal texture atlas.

In pre-alpha, that referred to a back copy of screen buffer. In the final, the scratch buffer reuses the `prim_objects` memory. `prim_objects` are basically the 3D objects for vehicles.

So, that texture shows bitmap interpetation of vehicle data.

We probably want to make that buffer usable, somehow. Though I don't know how, yet.
We currently can dislodge that buffer from vehicle data, and allocate separate memory for it - that part is already remade.
But what would we put there?

**Reply by Moburma** [reply]
I personally just wouldn't bother with the mirror stuff if it's not there, I'd just substitute it for a regular texture or something.

I was quite ill so didn't do much, but today finally got a nice first draft of the level made:

https://filebin.net/kds6g29n6drxkuph

I looked at the original and boiled it down to several things:

1. One Briefcase man walks in a kind of D shape around the traffic island
2. All white dress women in the level are set to follow him
3. the Spiders don't really seem hostile, in fact I'm not sure if they even can fight back. They do get into combat sometimes and die, so I just left them neutral for now as they are obviously combat beasts in the final game code
4. As far as I can see, all the zealots, agents, and one group of punks move to a certain area to have a big fight. They last long enough that it looks interesting, but as it is now, I set it to an area limited wander command, Unfortunately, because they must all go through the tight bottleneck of the bridge (and can't just wander across the water like they did in the original), this means it ends up as an all agent moshpit, with everyone else getting trounced, and zealots barely even getting near
5. The other group of punks seem to just be set to wander
6. Briefcase civs also wander

In the demo it seems some of the groups/individuals DO occasionally pick fights with the player, but I'm not sure how to do this as it seems somewhat random. The final game is far more all or nothing about this, so I just left all groups neutral to the player for now. The idea is more to observe than play, anyway.

Works fairly well, but some kind of massive rebalancing is needed so the battle isn't so massively one sided. The agents also seem faster than all other groups, so that also means they are all there first. Also in the demo they all seem to have loads of health as the fight goes on for an age.

**Reply by mefistotelis** [reply]
Yeah, that looks nice.

>  I'd just substitute it for a regular texture or something.

After disabling the flag 0x2, the faces go back to their normal look. This is what I'm doing, for now. Adding a proper feature there could be made when we have a proper map editor capable of utilizing such features.

>  I set it to an area limited wander command

I wonder if we can avoid them going in perfect circles by using two different time-limited wander command. Like, go to that area for 200 turns, then go randomly anywhere for 200 turns, then repeat.

Then how fast each group goes there can be controlled by changing the turn amounts.

> in the demo they all seem to have loads of health as the fight goes on for an age

You can check the original values in the level I dumped:
[C013L002.DAT-mefisto.md](https://github.com/swfans/swars/files/14320540/C013L002.DAT-mefisto.md)

My perception was that health was normal, just:
* all people seem not be all-in on the attacks, as you mentioned - they shoot one round, then move, then shoot again or select another target..
* some weapons seem to deal far less damage - like every weapon was an UZI with different shot animation

But increasing health may still be a good way to rebalance.

Also interesting - in the map I dumped, it looks like player agents have only lasers in hands - other weapons appear on them when you switch them, but they don't yet have the pockets to get them from.

**Comment by Moburma** [comment]
Another version:
https://filebin.net/85m6lugpu2ykk6vg

A lot of trial and error. I took away the stupid in-area wandering. This time I made everyone neutral, and made them all run to the "battle" area, which then gives them a kill all group command, with one of their two enemy groups listed (alternating between each member of the group, so one zealot has punks, one has agents, etc). This works reasonably well, looks less stupid and you actually get something of a battle.

I modded up the zealots and they now usually win, mainly due to them eventually arriving in a group when the agents are spread thin. Before this they would get trounced. I also removed the lasers from the spiders, what was happening was a zealot would graze one with his electron mace at the start, and then they would completely eradicate all zealots on the map in seconds.

Punks get completely trashed at the moment, so I guess the obvious next step is to pump them up to the level of the zealots. I replaced all the main NPC types with standard values for health etc, rather than the pre alpha ones, as they were just no good for the real game.

**Comment by Moburma** [comment]
Think I'll give up for now, this works quite well:

https://filebin.net/u1ploljz144hjo76

I tried to beef up the unguided, but this actually had the opposite effect. As there are fewer of them and they start off next to zealots, it just meant they mostly wiped each other out, and then the army of agents would finish them off in seconds. So I just left the unguided weak and zealots pretty strong. Had one session where only 1 zealot was left and no agents.

Might upload this to the levels repository in a while.

**Reply by mefistotelis** [reply]
Yes, please do.

I'm considering also adding the DAT map load ability to command line - ie. if MAD file does not exist, then try DAT with the same number.
The debug modes should also be enabled from command line...
